### PR TITLE
Speed up data in

### PR DESCRIPTION
**Pull request by alan-stokes**
improves on SDP by at least a factor of 2 for Ethernet connected chips and up to 22 times for none ethernet connected chips. 

links to other pull requests that need to be merged in at the same time
- [ ] spynnaker8: https://github.com/SpiNNakerManchester/sPyNNaker8/pull/123
- [ ] spynnaker: https://github.com/SpiNNakerManchester/sPyNNaker/pull/482
- [ ] FrontEndCommon: https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/295
- [x] SpiNNMachine https://github.com/SpiNNakerManchester/SpiNNMachine/pull/66
- [ ] GFE:  https://github.com/SpiNNakerManchester/SpiNNakerGraphFrontEnd/pull/79
- [x] PACMAN: https://github.com/SpiNNakerManchester/PACMAN/pull/152

### CODE REVIEW

**Comment by dkfellows** [comment]
This PR cannot be reviewed until it is no longer conflicting with `master`.

**Comment by alan-stokes** [comment]
no idea what ya on about.......did two merges, and neither conflicted......i say pycharm smartness beats github here

File: spinnman/model/executable_targets.py [hunk 1]
```diff
@@ -9,18 +10,34 @@ class ExecutableTargets(object):
...
+    )
+
     def __init__(self):
         self._targets = dict()
         self._total_processors = 0
         self._all_core_subsets = CoreSubsets()
+        self._binary_to_executable_types_map = defaultdict(set)
 
-    def add_subsets(self, binary, subsets):
+    def add_subsets(self, binary, subsets, executable_type):
```
Comment (dkfellows) [inline spinnman/model/executable_targets.py]: Add a default for `executable_type` (to the most common form in our users' code?) to reduce the possibility of incompatibilities.

File: spinnman/model/executable_targets.py [hunk 2]
```diff
@@ -32,21 +49,39 @@ def add_subsets(self, binary, subsets):
                 self._total_processors += 1
                 self._all_core_subsets.add_processor(subset.x, subset.y, p)
 
-    def add_processor(self, binary, chip_x, chip_y, chip_p):
+        # add to binary to executable type
+        self._binary_to_executable_types_map[executable_type].add(binary)
+
+    def add_processor(self, binary, chip_x, chip_y, chip_p, executable_type):
```
Comment (dkfellows) [inline spinnman/model/executable_targets.py]: Add a default for `executable_type` (to the most common form in our users' code?) to reduce the possibility of incompatibilities.

File: spinnman/utilities/io/memory_io.py [hunk 3]
```diff
@@ -197,7 +208,8 @@ class MemoryIO(AbstractIO):
     # ensure they are written before a read occurs
     __write_cache__ = dict()
 
-    def __init__(self, transceiver, x, y, start_address, end_address):
+    def __init__(self, transceiver, x, y, memory_write_function, start_address,
+                 end_address):
```
Comment (dkfellows) [inline spinnman/utilities/io/memory_io.py]: It's a shame that we've got an incompatible API change here.

Comment (alan-stokes) [reply spinnman/utilities/io/memory_io.py]: worth thinking about, currently safe though, as all uses have been covered...... what would be declaire as default? data in or sdp?

File: spinnman/model/executable_targets.py (hunk 1 above)
Comment (alan-stokes) [reply spinnman/model/executable_targets.py]: uses_simulation?

Comment (alan-stokes) [reply spinnman/model/executable_targets.py]: to do this, ill need to move the executable_type from FEC to....... spinnMan. not sure if we want to do this or not, as thats a much more substantial change, and not sure if it has context elsewhere than FEC really.

File: spinnman/model/executable_targets.py (hunk 2 above)
Comment (alan-stokes) [reply spinnman/model/executable_targets.py]: as above

**Comment by dkfellows** [comment]
I really dislike the way this introduces incompatibilities and talks about executable types without the types actually being defined.

File: spinnman/utilities/io/memory_io.py (hunk 3 above)
Comment (dkfellows) [reply spinnman/utilities/io/memory_io.py]: `transceiver.write_memory` is the backward-compatible approach. (That forces putting the parameter at the end, but that's OK.)

File: spinnman/processes/abstract_process.py [hunk 4]
```diff
@@ -11,37 +12,45 @@ class AbstractProcess(object):
...
-        self._error_request = request
-        self._exception = exception
-        self._traceback = tb
+        self._error_requests.append(request)
+        self._exceptions.append(exception)
+        self._tracebacks.append(tb)
 
     def is_error(self):
-        return self._exception is not None
+        return self._exceptions is not None
```
Comment (dkfellows) [inline spinnman/processes/abstract_process.py]: Well that's plain wrong. The `_exceptions` is never `None`, so there's always an error.

File: spinnman/transceiver.py [hunk 5]
```diff
@@ -663,6 +663,8 @@ def _update_machine(self):
         # Get the coordinates of the boot chip
         version_info = self.get_scamp_version()
 
+        time.sleep(5)
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinnman/transceiver.py]: yep. bodge for 3 bard toriod. mentioned numerious times. no real fix has come along yet. your welcome to do whatever you want with it. if it disappears. itll reappear in each new branch till the problem is fixed.

[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinnman/transceiver.py]: rowley is currently on the vcase, given my flagging it to him

[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinnman/transceiver.py]: fixed due to rowley work on master spinnman. removed accordngly

**Comment by alan-stokes** [comment]
ready for merging now me thinks. want to do a final eye ball donal?

File: spinnman/model/executable_targets.py (hunk 1 above)
Comment (rowleya) [reply spinnman/model/executable_targets.py]: So the issue here is that this now relies, though indirectly, on functionality in FEC i.e. the concept of an executable type.  This doesn't really have meaning in SpiNNMan that I can think of, other than for your filtering here.  It seems this part of the functionality might need a bit of a rethink.

**Comment by dkfellows** [comment]
It seems to me that the only real issue left is that executable types are an unknown concept to SpiNNMan. The code doesn't require that it knows… but it feels wrong nonetheless to rely on the undeclared functionality of a class that isn't currently visible.

**Comment by rowleya** [comment]
Agreed - I think the binary to executable type map could be kept outside of here.  If an extension object needs to be made in FEC that would also work.

**Comment by alan-stokes** [comment]
can this branch be deleted now? given donals revival of it?

**Pull request closed without merging**